src/coreclr/System.Private.CoreLib/src/System/Reflection/MdImport.cs

```diff
@@ -298,117 +355,126 @@ public void EnumEvents(int mdTypeDef, out MetadataEnumResult result)
             Enum(MetadataTokenType.Event, mdTypeDef, out result);
         }
 
+        #region FCalls
         [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern string? _GetDefaultValue(IntPtr scope, int mdToken, out long value, out int length, out int corElementType);
-        public string? GetDefaultValue(int mdToken, out long value, out int length, out CorElementType corElementType)
+        private static extern unsafe int GetDefaultValue(
+            IntPtr scope,
+            int mdToken,
+            out long value,
+            out byte* stringMetadataEncoding,
+            out int length,
+            out int corElementType);
+
+        public unsafe string? GetDefaultValue(int mdToken, out long value, out int length, out CorElementType corElementType)
         {
-            string? stringVal = _GetDefaultValue(m_metadataImport2, mdToken, out value, out length, out int _corElementType);
-            corElementType = (CorElementType)_corElementType;
-            return stringVal;
+            ThrowBadImageExceptionForHR(GetDefaultValue(m_metadataImport2, mdToken, out value, out byte* stringMetadataEncoding, out length, out int corElementTypeRaw));
+
+            corElementType = (CorElementType)corElementTypeRaw;
+
+            if (corElementType is CorElementType.ELEMENT_TYPE_STRING
+                && stringMetadataEncoding != null)
+            {
+                // Metadata encoding is always UTF-16LE.
+                return Text.Encoding.Unicode.GetString(stringMetadataEncoding, length);
```

Comment:
Should this be switched to `new string(` as well?

Reply:
I don't think so, but defer to your understanding of the issue. For `GetUserString()`, encoding binary goo makes sense, but for `GetDefaultValue()`, the constraints seem a bit more efforcible. I could be misunderstanding the scope of the problem though. I defer to what you think is best here based on historical use.

Reply:
@jkotas Unless you have a strong preference here, I am inclined to leave this.

1) The `#US` heap is where lots of goo seems to be encoded. There also seem to be mechanism that permit "unnatural" uses of strings.

2) The `GetDefaultValue()` API is going to look into the Constant (`0x0B`) table and seems to have a much clearer and defined intent.

Based on these observations I am advocating we become more restrictive with the `GetDefaultValue()` API and keep the permissive behavior for `GetUserString()`. We can easily change this if needed.

Reply:
From the C# user point of view, these are all identically looking string literals. Consider this program:
```
class Test
{
    static void Main()
    {
        Console.WriteLine((int)(((string)typeof(Test).GetMethod("M").GetParameters()[0].DefaultValue)[0]));
        M();
    }

    public static void M(string mangledKey = "\uD800\u1111") { Console.WriteLine((int)(mangledKey[0])); }
}
```
This prints two same numbers before this change and two different numbers after this change. It would be very hard to explain why the first character of this string literal is different, only in .NET 9+ and only when inspected by reflection. It looks like a needless breaking change to me. It is corner case, but the behavior change still fits the definition of a breaking change.

Reply:
>  only in .NET 9+ and only when inspected by reflection. It looks like a needless breaking change to me.

Agree. Making reflection do unexpected and difficult to predict things is not acceptable. Will change. Thanks.

Reply:
> This prints two same numbers before this change and two different numbers after this change.

FWIW, AOT returns different numbers before this change (even in .NET 8): 65533 and 55296.

Reply:
That sounds like something we should root cause. File an issue?

Reply:
https://github.com/dotnet/runtime/issues/101332